rust-cbindgen: fix test failures due to env vars

bryango opened the pull request:
... specifically, it's the `CARGO_BUILD_TARGET` env var introduced in commit: 683f97e3781356b3e8251d0ab5e5906c4cdd4ded. See: https://github.com/NixOS/nixpkgs/pull/298108.

Targeted `master` but tested against a cherry-picked 683f97e3781356b3e8251d0ab5e5906c4cdd4ded (which is currently `staging`).

Upstream PR is linked in the code; not linking here to avoid excessive notifications to upstream.
cc @trofi for testing :rocket: 

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [x] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

trofi commented:
It fixes `rust-cbindgen` build on `staging` for `x86_64-linux`. Thank you!

paparodeo commented on pkgs/development/tools/rust/cbindgen/default.nix:
```diff
@@ -23,6 +24,16 @@ rustPlatform.buildRustPackage rec {
+      # open PR: https://github.com/mozilla/cbindgen/pull/1010
+      # see also: https://github.com/NixOS/nixpkgs/pull/298108
+      name = "1010-fix-test-failures-due-to-CARGO_BUILD_TARGET";
+      url = "https://patch-diff.githubusercontent.com/raw/mozilla/cbindgen/pull/1010.patch";
```
this needs to be a stable url, like: `https://github.com/mozilla/cbindgen/commit/f8efe01d7f23c56b593ec686f0b430e6d47404e0.patch?full_index=1`, and the other commit, but also needs to be merged in the tree otherwise the commits can get garbage collected -- so with unmerged PRs we are just vendoring the patch rather than using `fetchpatch`.

bryango replied:
Good idea! I will wait till next week and see if it can get merged upstream. They are pretty responsive.

> stable url, like: `https://github.com/mozilla/cbindgen/commit/f8efe01d7f23c56b593ec686f0b430e6d47404e0.patch?full_index=1`

I have heard this before and will do so once upstream has responded. But just out of curiosity: how might `https://patch-diff.githubusercontent.com/raw/*/*/pull/*.patch` differ from the stable URL? Maybe the hash will change?

paparodeo replied:
` https://github.com/mozilla/cbindgen/pull/1010.patch` will change when the PR is updated, thus creating hash mismatch and breaking the package.

bryango commented:
I didn't expect this to be a mass rebuild :skull: I will wait for ofborg to finish first and retarget `staging`... For now I will mark this PR as draft just to be safe.

bryango committed:
rust-cbindgen: fix test failures due to env vars

... specifically, it's the `CARGO_BUILD_TARGET` env var introduced in
commit: 683f97e3781356b3e8251d0ab5e5906c4cdd4ded

bryango commented:
Since we already have a working patch (pending upstream), I think it would be better to vendor it (instead of disabling all the tests as is currently implemented in #348935). The patch is small and straightforward so it should be okay. What do you think? cc @alyssais @fabianhjr

emilazy reviewed (approved):
I’m okay with the patch; even if upstream wants changes it’s “only” tests that we’re not running without them anyway. Thanks!

bryango commented:
There is no cache available yet so I cannot test this on top of `staging-next`; it does build fine on top of master. Please ping me if things blow up :laughing:

emilazy commented:
Don’t worry, something always blows up.
